File: oxidize-golang/internal/api/schema.go
package api

import (
	"encoding/json"
	"fmt"
	"strings"
)

type ResponseFormatType string

const (
	ResponseFormatText       ResponseFormatType = "text"
	ResponseFormatJSONObject ResponseFormatType = "json_object"
	ResponseFormatJSONSchema ResponseFormatType = "json_schema"
)

type ChatCompletionRequest struct {
	Model               string          `json:"model"`
	Messages            []ChatMessage   `json:"messages"`
	ResponseFormat      *ResponseFormat `json:"response_format,omitempty"`
	GuidedJSON          json.RawMessage `json:"guided_json,omitempty"`
	JSONSchema          json.RawMessage `json:"json_schema,omitempty"`
	GuidedRegex         string          `json:"guided_regex,omitempty"`
	GuidedChoice        []string        `json:"guided_choice,omitempty"`
	Stream              bool            `json:"stream,omitempty"`
	MaxTokens           *int            `json:"max_tokens,omitempty"`
	MaxCompletionTokens *int            `json:"max_completion_tokens,omitempty"`
	Temperature         *float64        `json:"temperature,omitempty"`
	TopP                *float64        `json:"top_p,omitempty"`
	TopK                *int            `json:"top_k,omitempty"`
	MinP                *float64        `json:"min_p,omitempty"`
	TypicalP            *float64        `json:"typical_p,omitempty"`
	TailFreeZ           *float64        `json:"tail_free_z,omitempty"`
	Stop                StopSequences   `json:"stop,omitempty"`
	Seed                *uint64         `json:"seed,omitempty"`
	N                   *int            `json:"n,omitempty"`
	BestOf              *int            `json:"best_of,omitempty"`
}

type ChatMessage struct {
	Role    string         `json:"role"`
	Content MessageContent `json:"content"`
	Images  []string       `json:"images,omitempty"`
}

type MessageContent struct {
	text  string
	parts []ContentPart
	mode  messageContentMode
}

type messageContentMode uint8

const (
	messageContentText messageContentMode = iota
	messageContentParts
)

type ContentPart struct {
	Type     string           `json:"type"`
	Text     string           `json:"text,omitempty"`
	ImageURL *ContentImageURL `json:"image_url,omitempty"`
}

type ContentImageURL struct {
	URL    string `json:"url"`
	Detail string `json:"detail,omitempty"`
}

func NewMessageContentText(text string) MessageContent {
	return MessageContent{text: text, mode: messageContentText}
}

func NewMessageContentParts(parts []ContentPart) MessageContent {
	return MessageContent{parts: append([]ContentPart(nil), parts...), mode: messageContentParts}
}

func (c MessageContent) Text() string {
	if c.mode != messageContentParts {
		return c.text
	}
	var b strings.Builder
	for _, part := range c.parts {
		if part.Type == "text" {
			b.WriteString(part.Text)
		}
	}
	return b.String()
}

func (c MessageContent) Parts() []ContentPart {
	return append([]ContentPart(nil), c.parts...)
}

func (c MessageContent) MarshalJSON() ([]byte, error) {
	if c.mode == messageContentParts {
		return json.Marshal(c.parts)
	}
	return json.Marshal(c.text)
}

func (c *MessageContent) UnmarshalJSON(data []byte) error {
	var text string
	if err := json.Unmarshal(data, &text); err == nil {
		*c = NewMessageContentText(text)
		return nil
	}
	var parts []ContentPart
	if err := json.Unmarshal(data, &parts); err == nil {
		for _, part := range parts {
			switch part.Type {
			case "text":
			case "image_url":
				if part.ImageURL == nil || strings.TrimSpace(part.ImageURL.URL) == "" {
					return fmt.Errorf("decode content parts")
				}
			default:
				return fmt.Errorf("decode content parts")
			}
		}
		*c = NewMessageContentParts(parts)
		return nil
	}
	return fmt.Errorf("decode chat message content")
}

// MaxTokensOr returns max_tokens or max_completion_tokens, else defaultMax.
func (r ChatCompletionRequest) MaxTokensOr(defaultMax int) int {
	if r.MaxCompletionTokens != nil && *r.MaxCompletionTokens > 0 {
		return *r.MaxCompletionTokens
	}
	if r.MaxTokens != nil && *r.MaxTokens > 0 {
		return *r.MaxTokens
	}
	return defaultMax
}

// FirstUserMessage returns the last user message text in the chat.
func (r ChatCompletionRequest) FirstUserMessage() string {
	for i := len(r.Messages) - 1; i >= 0; i-- {
		if r.Messages[i].Role == "user" {
			return r.Messages[i].Content.Text()
		}
	}
	if len(r.Messages) > 0 {
		return r.Messages[len(r.Messages)-1].Content.Text()
	}
	return ""
}

// MaxTokensOr returns max_tokens or defaultMax when unset.
func (r CompletionRequest) MaxTokensOr(defaultMax int) int {
	if r.MaxTokens != nil && *r.MaxTokens > 0 {
		return *r.MaxTokens
	}
	return defaultMax
}

type CompletionRequest struct {
	Model          string          `json:"model"`
	Prompt         string          `json:"prompt"`
	ResponseFormat *ResponseFormat `json:"response_format,omitempty"`
	GuidedJSON     json.RawMessage `json:"guided_json,omitempty"`
	JSONSchema     json.RawMessage `json:"json_schema,omitempty"`
	GuidedRegex    string          `json:"guided_regex,omitempty"`
	GuidedChoice   []string        `json:"guided_choice,omitempty"`
	Stream         bool            `json:"stream,omitempty"`
	MaxTokens      *int            `json:"max_tokens,omitempty"`
	Temperature    *float64        `json:"temperature,omitempty"`
	TopP           *float64        `json:"top_p,omitempty"`
	TopK           *int            `json:"top_k,omitempty"`
	MinP           *float64        `json:"min_p,omitempty"`
	TypicalP       *float64        `json:"typical_p,omitempty"`
	TailFreeZ      *float64        `json:"tail_free_z,omitempty"`
	Stop           StopSequences   `json:"stop,omitempty"`
	Seed           *uint64         `json:"seed,omitempty"`
	Echo           bool            `json:"echo,omitempty"`
	N              *int            `json:"n,omitempty"`
	BestOf         *int            `json:"best_of,omitempty"`
}

type EmbeddingsRequest struct {
	Model string          `json:"model"`
	Input json.RawMessage `json:"input"`
}

type ResponseFormat struct {
	Type       ResponseFormatType `json:"type"`
	JSONSchema json.RawMessage    `json:"json_schema,omitempty"`
}

func (f ResponseFormat) OutputText() string {
	switch f.Type {
	case ResponseFormatJSONObject, ResponseFormatJSONSchema:
		return "{}"
	default:
		return ""
	}
}

type StopSequences struct {
	values []string
}

func (s *StopSequences) UnmarshalJSON(data []byte) error {
	var single string
	if err := json.Unmarshal(data, &single); err == nil {
		s.values = []string{single}
		return nil
	}
	var many []string
	if err := json.Unmarshal(data, &many); err == nil {
		s.values = append([]string(nil), many...)
		return nil
	}
	return fmt.Errorf("decode stop sequences")
}

func (s StopSequences) Values() []string {
	return append([]string(nil), s.values...)
}

File: oxidize-python/oxidize_python/internal/api/schema.py
"""OpenAI API request schema mirroring oxidize-golang/internal/api/schema.go."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any


class ResponseFormatType(StrEnum):
    TEXT = "text"
    JSON_OBJECT = "json_object"
    JSON_SCHEMA = "json_schema"


@dataclass
class ResponseFormat:
    type: ResponseFormatType = ResponseFormatType.TEXT
    schema: dict[str, Any] | None = None

    def output_text(self) -> str:
        if self.type == ResponseFormatType.JSON_OBJECT:
            return "{}"
        return ""


@dataclass
class ContentImageURL:
    url: str
    detail: str = ""


@dataclass
class ContentPart:
    type: str
    text: str = ""
    image_url: ContentImageURL | None = None


@dataclass
class MessageContent:
    text: str = ""
    parts: list[ContentPart] = field(default_factory=list)
    mode: str = "text"

    @staticmethod
    def from_text(text: str) -> MessageContent:
        return MessageContent(text=text, mode="text")

    @staticmethod
    def from_parts(parts: list[ContentPart]) -> MessageContent:
        return MessageContent(parts=list(parts), mode="parts")

    def text_value(self) -> str:
        if self.mode != "parts":
            return self.text
        return "".join(p.text for p in self.parts if p.type == "text")


@dataclass
class ChatMessage:
    role: str
    content: MessageContent
    images: list[str] = field(default_factory=list)


@dataclass
class ChatCompletionRequest:
    model: str = ""
    messages: list[ChatMessage] = field(default_factory=list)
    response_format: ResponseFormat | None = None
    guided_json: bytes | None = None
    json_schema: bytes | None = None
    guided_regex: str = ""
    guided_choice: list[str] = field(default_factory=list)
    stream: bool = False
    max_tokens: int | None = None
    max_completion_tokens: int | None = None
    temperature: float | None = None
    top_p: float | None = None
    top_k: int | None = None
    min_p: float | None = None
    typical_p: float | None = None
    tail_free_z: float | None = None
    stop: list[str] | list[int] | None = None
    seed: int | None = None
    n: int | None = None
    best_of: int | None = None

    @classmethod
    def from_json(cls, raw: bytes | str) -> ChatCompletionRequest:
        data = json.loads(raw) if isinstance(raw, (bytes, str)) else raw
        messages: list[ChatMessage] = []
        for m in data.get("messages", []):
            content = m.get("content", "")
            if isinstance(content, str):
                mc = MessageContent.from_text(content)
            else:
                parts = [
                    ContentPart(
                        type=p.get("type", "text"),
                        text=p.get("text", ""),
                        image_url=ContentImageURL(**p["image_url"])
                        if "image_url" in p
                        else None,
                    )
                    for p in content
                ]
                mc = MessageContent.from_parts(parts)
            messages.append(ChatMessage(role=m.get("role", "user"), content=mc))
        return cls(model=data.get("model", ""), messages=messages, stream=data.get("stream", False))
